curta/src/chip/ec/weierstrass/projective.rs

```diff
@@ -0,0 +1,310 @@
+//! Gadgets for SW elliptic curve projective points
+use core::marker::PhantomData;
+use core::ops::{Add, Mul};
+
+use num::{BigUint, Zero};
+use plonky2::field::types::PrimeField64;
+
+use super::WeierstrassParameter;
+use crate::chip::builder::AirBuilder;
+use crate::chip::ec::EllipticCurveParameters;
+use crate::chip::field::parameters::FieldParameters;
+use crate::chip::field::register::FieldRegister;
+use crate::chip::register::Register;
+use crate::chip::trace::writer::TraceWriter;
+use crate::chip::utils::{biguint_to_bits_le, field_limbs_to_biguint};
+use crate::chip::AirParameters;
+use crate::polynomial::to_u16_le_limbs_polynomial;
+
+#[derive(Debug, Clone)]
+/// short Weierstrass projective point
+pub struct SWProjectivePoint<E: EllipticCurveParameters> {
+    /// x coordinate
+    pub x: BigUint,
+    /// y coordinate
+    pub y: BigUint,
+    /// z coordinate
+    pub z: BigUint,
+    _marker: std::marker::PhantomData<E>,
+}
+
+impl<E: EllipticCurveParameters> SWProjectivePoint<E> {
+    /// Create a new SW affine point
+    pub fn new(x: BigUint, y: BigUint, z: BigUint) -> Self {
+        Self {
+            x,
+            y,
+            z,
+            _marker: std::marker::PhantomData,
+        }
+    }
+}
+
+impl<E: EllipticCurveParameters> Eq for SWProjectivePoint<E> {}
+impl<E: EllipticCurveParameters> PartialEq for SWProjectivePoint<E> {
+    fn eq(&self, other: &Self) -> bool {
+        if self.z.is_zero() {
+            return other.z.is_zero();
+        }
+        let modulus = &E::BaseField::modulus();
+        let zz1 = (&self.z * &self.z) % modulus;
+        let zzz1 = (&zz1 * &self.z) % modulus;
+        let zz2 = (&other.z * &other.z) % modulus;
+        let zzz2 = (&zz2 * &other.z) % modulus;
+        ((&self.x * zz2 % modulus) == (&other.x * zz1 % modulus))
+            && ((&self.y * zzz2 % modulus) == (&other.y * zzz1 % modulus))
+    }
+}
+
+impl<E: WeierstrassParameter> SWProjectivePoint<E> {
+    fn scalar_mul(&self, scalar: &BigUint) -> Self {
+        let mut result = E::neutral();
+        let mut temp = self.clone();
+        let bits = biguint_to_bits_le(scalar, E::nb_scalar_bits());
+        for bit in bits {
+            if bit {
+                result = &result + &temp;
+            }
+            temp = &temp + &temp;
+        }
+        result
+    }
+}
+
+impl<E: WeierstrassParameter> Add<&SWProjectivePoint<E>> for &SWProjectivePoint<E> {
+    type Output = SWProjectivePoint<E>;
+
+    fn add(self, other: &SWProjectivePoint<E>) -> SWProjectivePoint<E> {
```

Comment:
great reference algorithm! but should add unit tests. (maybe cross-check with arkwork's implementation?)

I don't see any tests checking its correctness.